tokio/src/sync/broadcast.rs

```diff
@@ -0,0 +1,980 @@
+        slot.write.pos.with_mut(|ptr| unsafe { *ptr = pos });
+        slot.write.val.with_mut(|ptr| unsafe { *ptr = value });
+
+        // Set remaining receivers
+        slot.rem.store(rem, SeqCst);
+
+        // Release the slot lock
+        slot.lock.store(0, SeqCst);
+
+        // Notify waiting receivers
+        self.notify_rx();
+
+        Ok(())
+    }
+
+    fn notify_rx(&self) {
+        let mut curr = self.shared.wait_stack.swap(ptr::null_mut(), SeqCst) as *const WaitNode;
+
+        while !curr.is_null() {
+            let waiter = unsafe { Arc::from_raw(curr) };
+
+            // Update `curr` before toggling `queued` and waking
+            curr = waiter.next.with(|ptr| unsafe { *ptr });
+
+            // Unset queued
+            waiter.queued.store(false, SeqCst);
+
+            // Wake
+            waiter.waker.wake();
+        }
+    }
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Sender<T> {
+        let shared = self.shared.clone();
+        shared.num_tx.fetch_add(1, SeqCst);
+
+        Sender { shared }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        if 1 == self.shared.num_tx.fetch_sub(1, SeqCst) {
+            let _ = self.send2(None);
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Lock the next value if there is one.
+    ///
+    /// The caller is responsible for unlocking
+    fn recv_ref(&mut self, spin: bool) -> Result<RecvGuard<'_, T>, TryRecvError> {
+        let idx = (self.next & self.shared.mask as u64) as usize;
+
+        // The slot holding the next value to read
+        let slot = &self.shared.buffer[idx];
+
+        // Lock the slot
+        if !slot.try_rx_lock() {
+            if spin {
+                while !slot.try_rx_lock() {
+                    spin_loop_hint();
+                }
+            } else {
+                return Err(TryRecvError::Empty);
+            }
+        }
+
+        let guard = RecvGuard {
+            slot,
+            condvar: &self.shared.condvar,
+        };
+
+        if guard.pos() != self.next {
+            let pos = guard.pos();
+
+            guard.drop_no_rem_dec();
+
+            if pos.wrapping_add(self.shared.buffer.len() as u64) == self.next {
+                return Err(TryRecvError::Empty);
+            } else {
+                let tail = self.shared.tail.lock().unwrap();
+
+                // `tail.pos` points to the slot the **next** send writes to.
+                // Because a receiver is lagging, this slot also holds the
+                // oldest value. To make the positions match, we subtract the
+                // capacity.
+                self.next = tail.pos.wrapping_sub(self.shared.buffer.len() as u64);
+
+                return Err(TryRecvError::Lagged);
+            }
+        }
+
+        self.next = self.next.wrapping_add(1);
+
+        Ok(guard)
+    }
+}
+
+impl<T> Receiver<T>
+where
+    T: Clone,
+{
+    /// Attempts to return a pending value on this receiver without blocking.
```

Comment:
Nit: I wonder if the use of "blocking" here is potentially confusing to users — none of this code ever blocks the _thread_, right? Should this be
```suggestion
    /// Attempts to return a pending value on this receiver without yielding.
```

Reply:
I just removed the section :shrug: